seisidd/experiment.py

```diff
@@ -401,16 +544,845 @@ def scan_closure():
+        #   epics_put("1ide1:Kohzu_E_dnHsize.VAL", (($1)+0.1), 10) ##
+        #   epics_put("1ide1:Kohzu_E_upVsize.VAL", ($2), 10) ## VERT SIZE
+        #   epics_put("1ide1:Kohzu_E_dnVsize.VAL", ($2)+0.1, 10) ##
+        _beam_h_size    =   cfg['nf']['beamsize_h']
+        _beam_v_size    =   cfg['nf']['beamsize_v']
+        yield from bps.mv(beam.s1.h_size, _beam_h_size          )
+        yield from bps.mv(beam.s1.v_size, _beam_v_size          )
+        yield from bps.mv(beam.s2.h_size, _beam_h_size + 0.1    )       # add 0.1 following 1ID convention
+        yield from bps.mv(beam.s2.v_size, _beam_v_size + 0.1    )       # to safe guard the beam?
+
+        # set attenuation
+        _attenuation = cfg['nf']['attenuation']
+        yield from bps.mv(beam.att.att_level, _attenuation)
+
+        # check energy
+        # need to be clear what we want to do here
+        _energy_foil = cfg['nf']['energyfoil']
+        yield from bps.mv(beam.foil, _energy_foil)      # need to complete this part in beamline.py
+
+        # TODO:
+        #   set up FS controls
+        #   decide what to do with the focus lenses
+
+        #######################################
+        ## calculate slew speed for fly scan ##
+        #######################################
+        # https://github.com/decarlof/tomo2bm/blob/master/flir/libs/aps2bm_lib.py
+        # TODO: considering blue pixels, use 2BM code as ref
+        scan_time = (acquire_time+cfg['nf']['readout_time'])*n_projections
+        slew_speed = (angs.max() - angs.min())/scan_time
+        cfg['nf']['slew_speed'] = slew_speed
+        
+        # need to make sure that the sample out position is the same for both front and back
+        x0, z0 = stage.kx.position, stage.kz.position
+        dfx, dfz = cfg['nf']['sample_out_position']['kx'], cfg['nf']['sample_out_position']['kz']
+        rotang = np.radians(cfg['nf']['omega_end']-cfg['nf']['omega_start'])
+        rotm = np.array([[ np.cos(rotang), np.sin(rotang)],
+                            [-np.sin(rotang), np.cos(rotang)]])
+        dbxz = np.dot(rotm, np.array([dfx, dfz]))
+        dbx = dbxz[0] if abs(dbxz[0]) > 1e-8 else 0.0
+        dbz = dbxz[1] if abs(dbxz[1]) > 1e-8 else 0.0
+        # now put the value to dict
+        cfg['nf']['initial_kx']       = x0
+        cfg['nf']['initial_kz']       = z0
+        cfg['nf']['fronte_white_kx']  = x0 + dfx
+        cfg['nf']['fronte_white_kz']  = z0 + dfz
+        cfg['nf']['back_white_kx']    = x0 + dbx
+        cfg['nf']['back_white_kz']    = z0 + dbz
+        
+        ## Ideally, we set up the FS control once, then the FS will be controlled with
+        ## intended signals
+        
+        ############################
+        ## Near Field Volume Scan ##
+        ############################
+        n_layers   = cfg['nf']['volume']['n_layers']
+        ky_start   = cfg['nf']['volume']['ky_start']
+        ky_step    = cfg['nf']['volume']['ky_step']
+        if ky_step == 0:
```

Comment:
for repeating scans, there is already one plan stub available.  If we want to do multiple detector distances, the movement of the nf-det should be included in the scan plan.

Reply:
can you share the repeating scan stub from Bluesky?  I'll check if I can easily apply that instead of this homemade approach...